### DISCUSSION TITLE
Asking the Community: Handling Keyboard Arrow Keys, When Freeplane moves from 2 directional layouts (left, right) to 4 directional layouts (left, right, up, down)

### DISCUSSION THREAD

**Discussion by sifran-github** [freeplane/freeplane#989] [General]
## SUMMARY

I discussed with @dpolivaev about it, and we decided to ask the community. In this post I want to:
1. Show that there are issues with the current implementation of keyboard arrow keys in the new 4 directional layouts.
2. My suggested solution (I have changed my mind!).
3. Asking the community for help and suggestions.

Note: I will use L,R,U,D to describe  left,right,up,down.

## FIRST, LETS SEE THAT THERE IS A PROBLEM
This is the keyboard mapping on the latest version freeplane_bin-1.11.1-pre12
![image](https://user-images.githubusercontent.com/5730355/217275497-c5a9fd6c-3f50-491d-860f-54a145ef214b.png)
as you can see, no key will lead us to the next sibling.
It is an issue and needs to be solved. The only question is: how?

Another example:
![image](https://user-images.githubusercontent.com/5730355/217282552-84559fbe-00f8-46f9-bf00-7d9dc8b269b3.png)

The current implementation is confusing.

## MY PREVIOUS SUGGESTION (I CHANGED MY MIND FROM FUNCTION TO DIRECTION)
keys can be handled by FUNCTION or by DIRECTION. in case of FUNCTION, for example, the right key will always "go into the next child node". It will perform the same FUNCTION every time, regardless of the layout.

Previously, I thought that the correct solution for 4 directional layouts is to use keys by FUNCTION, But I changed my mind on this issue.

I noticed that even today, in 2 directional layouts, keys are behaving based on DIRECTION and their function is changing based on the current layout:

## TODAY, IN 2 DIRECTIONAL LAYOUTS, KEYS ARE BASED ON DIRECTION
Lets imagine that we have a map:
![image](https://user-images.githubusercontent.com/5730355/217266411-35ef24d0-fd48-47ed-bc02-ffcfa90a5a79.png)

Now lets imagine that we select a node. how the keys are responding?

![image](https://user-images.githubusercontent.com/5730355/217267964-bb97b9b3-410d-435c-bc09-25eafd4dd369.png)

SELECTED: RIGHT 1
L: ROOT
R: RIGHT 1-1 (or the previously selected sub-item)
U: [nothing]
D: RIGHT 2

SELECTED: LEFT 1
L: LEFT 1-1 (or the previously selected sub-item)
R: ROOT
U: [nothing]
D: LEFT 2

As you can see, keys left and right change their function based on the layout and therefore based on DIRECTION (on the right side, right key will be "go to child node"and on the left side, the same key will be "go to parent")

## WHAT TO DO WITH 4 DIRECTIONS?
I'm actually not really sure, but i do support keeping it by DIRECTION and changing keys based on the active layout.
The thing is, there are so many different new layouts... that it is very confusing for me to suggest a complete rules-based-system for every possible layout.

So, instead, I will try to add an example of a behavior that i think we should start with:

![image](https://user-images.githubusercontent.com/5730355/217284694-0b6066d5-2a0f-4556-b12e-6047ed1161d6.png)

ANOTHER ONE, SAME LAYOUT, FROM THE PARENT NODE:
![image](https://user-images.githubusercontent.com/5730355/217285756-c22eed4a-303c-4ab1-bca3-44f675d0b191.png)

## COMMUNITY, PLEASE COMMENT AND HELP
Freeplane Community: What do you think?
Do you like my proposal? Do you think that it is enough?
Do you have an idea for a complete rules-based-system for every possible layout?

THE MINDMAP FILE: [test_new_version_v2.zip](https://github.com/freeplane/freeplane/files/10677064/test_new_version_v2.zip). I have used version freeplane_bin-1.11.1-pre12 to edit the mind map. The file might be a bit different than the screenshots.

**Comment by sifran-github** [comment]
Bonus Question:

In the following layout, each node is a child of the previous node. there are no siblings:
![image](https://user-images.githubusercontent.com/5730355/217295234-2ad41fcb-2aca-418f-a111-4ea0761af5af.png)

CIRCLE is a parent of CIRCLE A
CIRCLE A is a parent of CIRCLE B
CIRCLE B is a parent of CIRCLE C
and so on.

There is a different layout for each node, to create a circular shape of nodes.

Question: what keys do you think that users SHOULD press in order to go from node CIRCLE to node CIRCLE F?
I emphasis the word should, because at the moment it is impossible to do so with the keyboard.

**Reply by sifran-github** [reply]
I would say the most intuitive answer is: D, R, U, L, L, D.
This answer is totally based on DIRECTION and it means that each layout has to know how to deal with the keyboard in a different way.
(all these different keyboard keys will activate the same function, "go to child node", based on the selected node's layout).

**Reply by bepolymathe** [reply]
> Question: what keys do you think that users SHOULD press in order to go from node CIRCLE to node CIRCLE F?

Ideally I would say UP... but my problem is different, I don't even understand how I go from CIRCLE C to CIRCLE D 🤔

**Comment by EdoFro** [comment]
> ## WHAT TO DO WITH 4 DIRECTIONS?
> I'm actually not really sure, but i do support keeping it by DIRECTION and changing keys based on the active layout. The thing is, there are so many different new layouts... that it is very confusing for me to suggest a complete rules-based-system for every possible layout.
> 
> So, instead, I will try to add an example of a behavior that i think we should start with:

Hi @sifran-github,
I think you have a good point here. I have no obvious solution for thies.
I just want to ask you to change the images and use the node Mid-R-2 as example, because I can't extrapolate your proposal for that node (or any node that is not directly under its parent, and I want to understand your proposal.

Thanks

edo

PD: The more I think about it, the more I believe that it should be a combination of DIRECTION and FUNCTION keys. (For example: arrows for direction (no matter relationship between nodes, just jumps to the nearest node in that direction), ESC goes to parent, '.' and ',' goes to previous or next node in same level, etc.)

PD2: This is a very important issue, because moving node with the keyboard uses the arrows in the same way.

**Reply by sifran-github** [reply]
Regarding MID-R-2 ... I will try to answer with my best suggestion:

![image](https://user-images.githubusercontent.com/5730355/217380419-999ccf6d-0b96-452c-9a2a-eedab71dd4d9.png)

(And if MID-L-2 and MID-L-3 dont exist.... then pressing Left should probably do nothing? or maybe jump to MID-L-1?)
what do you think about the suggestion?

I don't have answers to all the proposed layouts. There are many new layouts and two more dimensions to deal with.... this is why i am asking for help from the community.

---

On the other issue that you mentioned, are you suggesting that we should add more keys that are based on FUNCTIONs and provide the functionality itself in any layout? because as of today, ESC goes to root and , and . are just like pressing any letter, they start editing the node (in my settings anyway).
My response: maybe, yes. But it will require learning more keyboard shortcuts and might take keys that are already in use, so maybe its not as good. But... if the arrow keys are confusing, it might be helpful.

Community... we need your help...

**Comment by bepolymathe** [comment]
The situations are so different that I would tend to say that the closest and simplest DIRECTION solution is the best. A bit of a VIM move. A FUNCTION keys plus the DIRECTION as proposed by @EdoFro seems to me a good solution to complete the current hierarchical shift logic.

But the best visual move I've experienced is the one based on a hotkey that displays letters or numbers on top of the destinations and pressing the character makes you jump there (like vimium like solution).

Well I know that doesn't really answer your question but I'm still hesitating...

**Comment by bepolymathe** [comment]
> ANOTHER ONE, SAME LAYOUT, FROM THE PARENT NODE:

For this situation, the most intuitive for me would be that 
From note "MID" (as actual)
- UP --» CATEGORY 2
- LEFT --» MID-L-1
- RIGHT --» MID-R-1
- DOWN --» MID-R-1

But from "MID-R-1"
- UP --» MID (instead of LL3)
- LEFT --» MID-L-1 (instead of MID) 
- RIGHT --» MID-R-1-1 (as actual)
- DOWN --» MID-R-2 (instead of MID-R-1-1)

**Reply by sifran-github** [reply]
Regarding your suggestion from "MID": in this case, how can the user navigate to the siblings of MID? (LLL and RRR).
The user is kind of locked in side the MID tree and cannot simply navigate to these silbings only with they keyboard, and have use the mouse.
In my suggested solution for this layout the user can either go to the LLL/RRR siblings by using the left and right keys, or if he wants to jump to the children MID-R-1 or MID-L-1, he can first use the down key and then the left-right keys.

Regarding your suggestion from "MID-R-1": looks like we think about the same suggested keyboard response.

**Comment by sifran-github** [comment]
## Thinking about some rules for keys
Maybe the 4 keys (left,right,up,down) should perform these 4 different functions:
1. go to child node (the previously selected one, just like today).
2. go to parent.
3. go to next sibling.
4. go to previous sibling.

(and based on DIRECTION = the layout defines which key does what function)

Which also means that there should not be a situation where two different keys perform the same function, taking away some another functionality. You can see an example of such situation in the first picture of the original post, where two keys (right and down) perform the same function "go to child" but there is no key for the function "go to next sibling".

## But still, some new layouts are confusing.

Even with this rule, the MID layout is still new and confusing to solve.

Actually, now that i look at it again, even in my SUGGESTED solution, when the selected node is MID-R-2... there is no key that perform the "go to parent" functionality.
So, for this layout, from the central nodes, there will be no "go to parent"? I'm still not sure.

**Comment by dpolivaev** [comment]
I'd like to explain why this question is so challenging: we have only 4 directions but we actually need more.
Let me list them:

- two directions go to the next sibling and to the previous sibling
- two directions go to child nodes if whey are placed at the both sides
- one direction goes to the parent node.

In the following examples, S-nodes are children of the node called `Parent` and C-nodes are children of some S-nodes

If node siblings and its children are laid out so that the layout direction does not change, e.g. from top to bottom or from left to right, one of the two child sides points toward a parent which makes a collision, and the other away from the parent.

Here both siblings and their children are laid out left to right:
<img width="182" alt="AllLeftToRight" src="https://user-images.githubusercontent.com/1558052/217915049-ee99d7a5-396d-41bb-80a9-0eafc37ce738.png">

Here both siblings and their children are laid out top to bottom:
<img width="343" alt="AllTopToBottom" src="https://user-images.githubusercontent.com/1558052/217915266-24b2fe0c-2ef5-4a82-8074-c2be49dc6f9f.png">

Children C43 and C44 have the same direction as parent relative to their parent nodes S3 and S4

If node siblings and its children are laid out so that the layout direction changes, the children and the siblings of S-nodes have the same direction relative to the S-node:

<img width="260" alt="SiblingsLeftToRightChildrenTopToBottom" src="https://user-images.githubusercontent.com/1558052/217916219-ea375610-1a61-4c73-b2ba-e6778920668e.png">

<img width="263" alt="SiblingsTopToBottomChildrenLeftToRight" src="https://user-images.githubusercontent.com/1558052/217916258-ed95358e-fd8a-472f-b0a9-fcbfc8829891.png">

The problem is not limited to navigation, it is also relevant to moving nodes using control+arrow keys.

The current solution is not perfect. And I am looking forward to your suggestions. I would appreciate it if you could use my example in your answers, therefore I attach the mind map to this post.

[layoutExamples2.mm.zip](https://github.com/freeplane/freeplane/files/10700825/layoutExamples2.mm.zip)

**Reply by bepolymathe** [reply]
Maybe I'll say something stupid. What if we consider that all configurations where there are more than 4 directions imply that the parent node becomes a kind of temporary central node to which we return with the ESC key. Pressing ESC twice would return to the real central node or move up one level in the branch. 🤔 I try to "delete" a direction with the arrows...

**Reply by dpolivaev** [reply]
I can assign ESC to "select parent" instead of "select root".
But it does not solve the collisions from the last two examples where child nodes and siblings appear in the same direction.
And it does not solve moving nodes.

**Reply by bepolymathe** [reply]
You're right, it's not wonderful. But if we take the last example. 

If the UP arrow is free because it is replaced by the ESC key and the cursor is in S3 then the UP key could go to S2 (clockwise) and it's a bit like following the branch to go to S2. 

But to be coherent, it should be the opposite with the DOWN key when you are in S4. 

But this raises the problem of going from S4 (in the last example) to S2 (in the previous example). And it's not great for the moves...

**Reply by sifran-github** [reply]
Just a small note that ESC is not good for "go to parent" because its not what ESC represent.
I do not disagree with having special FUNCTION keys, but if we do, we need to choose logical keys.

**Comment by sifran-github** [comment]
I have an idea for a possible solution:

_**POSSIBLE SOLUTION**_
There will be two modes of moving:
1. **Arrow keys** (default): Navigate ONLY to siblings (not to children/parent), based on the selected node's PARENT LAYOUT for understanding where the siblings are. In case that there are no sibling in that direction, do action 2.
2. **KEY (ALT?) + Arrow keys**: Navigate ONLY to parent or children nodes, based on the PARENT LAYOUT node for understanding where the parent is, and the selected node's CURRENT LAYOUT for understanding where the children are. if the parent and children have the same direction (which is a very uncommon), go to parent. if there are no children/parent in this direction, do nothing (do not do action 1).

This is the entire idea and the rules to implement it.

_**NAVIGATION IMAGES**_
The solution implemented in the very complex mid layout:
![image](https://user-images.githubusercontent.com/5730355/218302279-97d240c1-e739-4038-a7e4-aec6b5ccd9af.png)

![image](https://user-images.githubusercontent.com/5730355/218301985-909a8ad1-9ff9-4b5d-a30d-449121cbf7d4.png)

in @dpolivaev 's map:
![image](https://user-images.githubusercontent.com/5730355/218302513-0398f341-d376-43f7-accf-b21bcf1e8d56.png)

_**NOTES:**_
Remember that in many layouts, pressing the arrow key alone will not find a sibling in that direction and the action will be action 2.

why siblings are the default? because if i compare it to a navigation inside a folder with sub folders, the default behavior is go go to siblings (folders on the same level) and then you do a special operation to go inside a folder (ENTER) or a special operation to go to the parent (ALT+UP ARROW).

ALT+arrows are controlling the zoom today. we can replace it with the keys of every browser today for zoom, Ctrl+ and Ctrl-. these keys are used today in FP to set the node's font size. I think that font size doesn't really need a shortcut by default (the user can set it if he likes).

Moving Nodes, note 1: I don't know how it will affect moving nodes, which i'm not sure how the algorithm works exactly. Do we need a special key, like i suggested here, for moving nodes as well? i don't know.

Moving Nodes, note 2: I actually changed my keys for moving nodes to ALT+SHIFT+Arrow keys, because some application are using it and i am kind of used to it. so if we do use ALT for moving, i'm not sure how it will affect my settings...

I am not sure how to navigate to/from free positioned nodes, but its an edge case, and we can look how Freeplane behaves today.

Please let me know what you think.

**Comment by BobLangill** [comment]
The topic states "when Freeplane  **moves** from 2 directional layouts" This implies changes to the ways we've been navigating, moving, and selecting nodes even if the map structure is a basic folding tree.  Will there be a "left/right mapping only" option we can select for an individual map that preserves the behaviors we're used to?

**Reply by dpolivaev** [reply]
Yes, you can check the latest preview 1.11.1_14

**Comment by sifran-github** [comment]
Any other comments from people in the community?

I think that this is an important issue.
Otherwise, in future releases, it will be almost impossible to navigate mind maps using the keyboard .

**Reply by Procra55** [reply]
Hallo, thank you for this invitation. It is exciting to see the rapid development of Freeplane, which I like very much. I use it since years for simple note taking and as a information database, and I appreciate the new possibilities for map layout.

I have a proposal for very intuitive navigation of the map tree by introducing a new graphical element: while traversing the map, select and highlight as well as the nodes the edge crossings and t-crossings.
This way you could choose with the arrow keys which path the "cursor" should follow, quite similar to Pac Man moving in his labyrinth in the old computer game.

Of course I have no idea how much work would be involved to change the existing code base of Freeplane to introduce this new behaviour.

Is this technically possible? And would anybody

**Reply by dpolivaev** [reply]
I have already reworked it. I hope I can release the preview this weekend.

**Comment by dpolivaev** [comment]
Preview 1.11.1_16 with new logic released

**Comment by sifran-github** [comment]
in Preview 1.11.1_16, this is the current logic:
![pre16_1](https://user-images.githubusercontent.com/5730355/223072289-9faa1e7d-8f4c-4544-8274-fb4b2dea00fc.png)
I think we should prefer "go to siblings" to "go to children". in this case, L and R should go to siblings and D should go to children (right now D does nothing).

![pre16_2](https://user-images.githubusercontent.com/5730355/223072297-41a9da24-2c57-4f26-b1d5-8291ce5eb7dd.png)
U is currently jumping to parent of parent! i think it should just jump just one level, to the parent.
L should go to the sibling node...

**Reply by dpolivaev** [reply]
Yes, I have not considered cases where child nodes are aligned "before parent" and "after parent" yet, you are right.
Next weekend I want to improve it.
For now, I suggest that you test all other alignments.

**Reply by sifran-github** [reply]
can you point out to the code in github of how the selection of the nodes is made, when pressing the keys?
If its simple, it might can help me to understand the general rules and have some suggestions.

**Reply by dpolivaev** [reply]
org.freeplane.view.swing.map.MapView.selectRelatedNode(SelectionDirection, boolean)

**Reply by sifran-github** [reply]
Forgive me if this is not 100% true, i tried to read the code and have some quick conclusions:

I think the bug of jumping to parent of parent is because there is a code that checks if the parent has requiredLayoutOrientation in the getVisibleSummarizedOrParentView function. i'm not sure why because any parent should be selected. maybe you decide to check the parent layout... but... this leads to the second point:

I think that the way to fix it is to change the logic a bit: at the moment the program (code at selectRelatedNode) chooses which node to go to based on the current selected node layout (selectedUsesHorizontalLayout) only.

I think the logic should be based on the selected node layout and the parent layout (not parent of parent and so on... just the closest parent). only then you can determine which direction should the selection move (and in my opinion, prefer siblings over children).

Unless you see a reason to jump to parent of parent of parent.... but i dont see any reason to do so in any case.

I hope what i wrote is true based on my understanding of the code.

**Reply by dpolivaev** [reply]
To discuss the logic we actually need not the code but example mind maps. There are situations where current logic makes sense. There are other situations.

Basically, I developed current logic based on https://github.com/freeplane/freeplane/files/10700825/layoutExamples2.mm.zip

**Reply by dpolivaev** [reply]
I am going to come back to Freeplane on Friday.

**Reply by sifran-github** [reply]
another quick bug from the layout you posted:
![image](https://user-images.githubusercontent.com/5730355/223090395-6019c536-55fe-49e4-b409-389dea57fc9d.png)
from this root, it jumps to the child of a child, two levels, instead of one.

**Reply by sifran-github** [reply]
to test, i suggest using also the map that i posted which contains more "exotic" layouts:
https://github.com/freeplane/freeplane/files/10677064/test_new_version_v2.zip
(link can be found in the original post as well)

**Reply by sifran-github** [reply]
"There are situations where current logic makes sense" - in which situations jumping more than one level makes sense? (for example jumping to a parent of a parent)

**Reply by dpolivaev** [reply]
> another quick bug from the layout you posted: ![image](https://user-images.githubusercontent.com/5730355/223090395-6019c536-55fe-49e4-b409-389dea57fc9d.png) from this root, it jumps to the child of a child, two levels, instead of one.

Actually not, here S4 is a child of Parent and C44 is a child of S4.

**Reply by sifran-github** [reply]
oh right. so this is not a bug then.

**Reply by dpolivaev** [reply]
Please check the new preview 1.11.1-17. The task is not easy. Your feedback and findings are appreciated.

**Comment by sifran-github** [comment]
OK i have tested 1.11.1-17 and the keyboard navigation feels much better!
I navigated the original testing map without any major issues, something that was just impossible to do before.
I also was able to easily navigate the circular part of the mind map in an intuitive way (impressive).

I still want to do an extensive test with more layouts combinations and more features testing like summary nodes etc', but I wont be able to do it in the next few days.

One small issue that i want to point out, but maybe its designed this way, is that in any "outline" layout, pressing Down will go to children and pressing Up will skip the children. I think that it will be better if pressing Down will go to the next sibling, since pressing Right is already going to the children, and since in my opinion siblings importance > children importance in most cases of the keyboard navigation.
![image](https://user-images.githubusercontent.com/5730355/225944864-55711898-0bbe-4b84-a326-ec904b86911f.png)

But anyway it is an amazing work. great improvement.
I hope to test it more soon.

**Reply by dpolivaev** [reply]
I improved navigation in outline view and navigation in filtered maps in preview 1.11.1_18

**Reply by sifran-github** [reply]
It seems to be fixed now. but after playing with the keys more and more, i'm having second thoughts about it.

I did hope to get more feedback about this specific navigation decision from the community.

I will explain the issue: Basically there are 3 options:
1. Like it was until now: that D goes children, U goes to sibling. I think that this is confusing anyway, and its good that we change it.
2. both U and D go to siblings (like it is in 1.11.1_18). I think it is a better than option 1.
![image](https://user-images.githubusercontent.com/5730355/226140584-4088e84a-3e5f-409b-9731-82b04df45d4d.png)
3. both U and D go to visible children to create a "smoother" navigation between all the outline levels (from MID-R-2, pressing U will go to MID-R-1-2, the last visible children of the previous siblings). SimpleMind uses this navigation decision. After playing with it for some time in SimpleMind, I actually do think that some people might want to have the navigation like this. It does feel smoother somehow.
![image](https://user-images.githubusercontent.com/5730355/226140563-9da7370d-3b7d-415a-af5a-b9750e355326.png)

Another reason why I'm starting to think that maybe 3 might be a better option is because sometimes pressing U and D will jump to the parent (do a "level switching"), and then if the user navigates naturally with UP and DOWN keys which should be smooth navigating between items, it can cause some confusion and unexpected "jumps". I will try to explain with a picture:
![image](https://user-images.githubusercontent.com/5730355/226141012-ecd69b47-742c-4a76-aa8b-adc3a5a58a51.png)
The user pressed D twice (so he jumped to a parent), and now he tries to go back and pressing U but he now already navigates in a whole new level. I think that if we do allow this "jump to the parent" by using U and D (which i think is OK), then this behavior might be confusing.

So @dpolivaev  i'm sorry about the confusion, but after testing it again, maybe option 3 is better than 2?
Other people want to try and comment?

**Reply by euu2021** [reply]
I think that, as much as possible, the behavior should be similar to the behavior of the classic layout. That said, it seems that option 2 is the most similar to the navigation behavior that we are used to and was tried and tested for years: pressing Up here goes to the sibling, and not to the nephew:

![image](https://user-images.githubusercontent.com/77707706/226143353-0397dadd-ad05-41d4-9acc-92afc0a80361.png)

Now, when there is no sibling above, should pressing up go to the uncle or to the cousin (like in the classic layout)?
![image](https://user-images.githubusercontent.com/77707706/226143508-ba117944-1783-4a50-81ed-8232da96f9c7.png)
 
If the parent is too wide, like in this example of yours, it is indeed visually weird if the selection goes to the cousin (at the second step):
![image](https://user-images.githubusercontent.com/77707706/226143603-edacc61a-7cb4-48fd-9cde-29e25610f90b.png)

But, I tend to prefer sticking with the classic behavior, anyway (ie, it goes to the cousin), for the sake of standardization.

**Reply by sifran-github** [reply]
I also thought that 2 is the best option, but... after actually playing with it and trying options 2 and 3 i tend to choose option 3 for all the outline layouts.

I suggest that you do the following:

- to test option 2: download freeplane preview 1.11.1_18, load the mindmap in my original post above, and play with it. I find some parts of the navigation still non-intuitive.
- to test option 3: use simplemind (using list layout or top down layout), or use the following demonstration that chatgpt and I wrote: https://jsfiddle.net/10hp3byq/ . only Up and Down are supported. it just shows the smoothness of it.

play with it, see how it feels.
I think that you will understand my reasoning for changing my mind, from option 2 to 3 (for outline views only).

**Reply by euu2021** [reply]
Ok, I tested the most famous outliners out there (they all have live demos in their sites): Workflowy; Dynalist; Roam Research; Logseq. All of them use option 3. So, it's better to not reinvent the wheel and just stick with it.

And this is probably the most visually intuitive option, anyway: pressing Up or Down goes to the visually closest line, although it may be logically distant. So, yes, I will go with option 3, too.

**Reply by dpolivaev** [reply]
Option 3 implemented in preview 1.11.1_19

**Reply by sifran-github** [reply]
Tested 1.11.1_19 with the mindmap in my original post, and it feels really smooth and intuitive.
Its the first time that i can say that i can easily navigate to each node using only the keyboard.

At least for this layout, its really good.
I will test more layouts in the upcoming week. If you have some mindmaps (or ideas for mindmap layouts) that i can test, please post them.

**Reply by dpolivaev** [reply]
I created preview 20 because I wanted to improve navigation to/from summary nodes

**Comment by EdoFro** [comment]
Hi all,
after following this discusion, I want to repeat a feature request expressed [here](https://github.com/freeplane/freeplane/discussions/960#discussioncomment-4898388) :

> to have the possibilty to open the map in a "right left only" layout that ignores all the top/down node's information, to work in the map as always, using the standard hotkeys

thanks,
edo

**Reply by EdoFro** [reply]
so the user can choose between:
- mindmap view
- outline view
- ‘left/right only’ mindmap view

**Reply by euu2021** [reply]
By chance, this suggestion made a lot of sense to me, today, because I took some time to test how the different layouts can be useful for my daily use, and noticed that they pretty much unlock a "Canvas" mode in FP. 

Basically, a Canvas mode is a special setting (like a playground) where the user can freely move nodes and branches around, without affecting the usual view of the map. Obsidian introduced it recently and it looks good.

Without the new layouts, this Canvas is too limited, because the branches have this boring fixed horizontal growing pattern. Now, it's different. And it's even better now that FP has the Jump In feature.

So, I will add another option to the ones that Edo listed above: Canvas view. In this mode, each node would have 3 properties that only exist inside this mode:

- node layout
- free positioned node
- node position

**Reply by dpolivaev** [reply]
Could you please share concrete examples of mind maps and concrete use cases? Abstract considerations do not deliver any insights which I need. I need to see the true picture in your perception, not just conclusions, because otherwise, I can not understand which actual problems and tasks you are trying to solve.

**Reply by euu2021** [reply]
Well, what I would like to replicate is the workflow that Niklas Luhmann did with physical paper, and is actually pretty simple: when I want to work on a problem, I collect ideas from several places of my knowledge base, temporarily put them on a "table", and start playing with them (combining, mixing, connecting, ordering, grouping, rewriting, improving etc).

A simplified step by step of what Luhman used to do:

- 1: As he reads books, he writes small notes and structures them in a tree like fashion. The tree is only logical, because, physically, the cards are in a simple sequence (but, this is a limitation of using physical paper, so, in FP, the tree is actually visible). The notes are like the nodes in FP.
- 2: At some point, he wants to write a paper about the Subject X
- 3: So, he goes through his notes, collecting the ones that may be useful
- 4: He puts all those notes in a table, and starts playing with them. The notes come from completely different contexts, so it's important to have a flexible space to play with the notes and combine them in a creative way
- 5: After finding a useful order in the notes, he writes the paper, based on that order
- 6: He puts the cards back to their original places


In FP, to do steps 3 and 4, I would go by cloning the nodes in a node called Playground. I clone the nodes, instead of copying, because I want to edit the cards (rewrite, connect, develop) while playing with them in the Playground. The Playground works well, now that FP has different layouts and the Jump In feature:

![image](https://user-images.githubusercontent.com/77707706/226213081-9f9e8dc3-410d-4db9-9b4c-c0c59f994d67.png)

But... the mess shows up when the user leaves the Playground and looks at the ideas branch. The changes in positioning and layout were carried out to there, too:

![image](https://user-images.githubusercontent.com/77707706/226213069-7d04bf97-27b2-4612-821f-c470e75b5da5.png)

What I suggest is that the user see this, instead, when changing from "Canvas" view back to the usual "Mind Map" view:
![image](https://user-images.githubusercontent.com/77707706/226215850-7dcf5a97-41e6-4683-bd22-0d73fb5e2993.png)

**Reply by euu2021** [reply]
[this video](https://www.youtube.com/watch?v=eHI-Szjpafk) shows how the Canvas work in Obsidian. It's interesting that she selects the nodes directly from the Canvas, and it's possible to do in FP with my Teleporter script set to clone the nodes.

**Reply by sifran-github** [reply]
@euu2021 :

1. Wouldnt it be easier for you to write a simple script "change all layouts of children nodes to auto" (revert their layouts to default) ? so after you finished with the playground, just go to the ideas parent node, apply the script, and everything will be aligned again.
2. Regarding canvas, i think that some parts of it can be done with floating nodes, but i like that Freeplane is very structured. Allowing too much "drawing freedom" will create a different program in my opinion. When i want to do "free drawing" i go to programs like Powerpoint and https://app.diagrams.net/. Maybe i missed the point but this is what i feel when watching the video that you posted.

**Reply by euu2021** [reply]
> Wouldnt it be easier for you to write a simple script "change all layouts of children nodes to auto" (revert their layouts to default) ? so after you finished with the playground, just go to the ideas parent node, apply the script, and everything will be aligned again.

Yes, I will probably start using it like this for simple situations. But, more complex projects require that the Playground stays there for some time, while life goes on outside of it.

> Regarding canvas, i think that some parts of it can be done with floating nodes, but i like that Freeplane is very structured. Allowing too much "drawing freedom" will create a different program in my opinion. When i want to do "free drawing" i go to programs like Powerpoint and https://app.diagrams.net/. Maybe i missed the point but this is what i feel when watching the video that you posted.

That's how I used to think until... yesterday. The new layouts may change the game here. There is structure, but now it is way more flexible, as different layouts can be combined.
The problem with those other apps is that they are missing the most important thing: my data. As I build all me knowledge base in FP, I would like to have an integrated setting where I can work on my data from different perspectives.

> Allowing too much "drawing freedom" will create a different program in my opinion.

Notice that I'm not asking for more drawing freedom. As I said above, I think the Playground is already working fine to play with my nodes. The problem is that it's not possible to create an environment "isolated" from the rest of the map.

**Reply by euu2021** [reply]
Ok, I think this can be done with scripts. The idea is:

- create an attribute line for each "perspective" that you may want to use
- the attribute values will contain a concatenated sequence of values for each of the elements that you want to isolate from the other "perspectives". For example, if the user wants to isolate "layout", "free positioned node", and position (horizontal and vertical shift), the value will be like "AUTO, TRUE, 23, -34"
- a listener automatically saves the current values in the attributes
- if the user activates "perspective 1" (choosing in a combobox in a GUI), the nodes are set to the values of the attribute line "perspective 1", and the listener records any change into this line
- when the user changes to "perspective 2", the nodes are set to the values of the attribute line "perspective 2", and the listener records any change into this line

Something cool about it is that the user can have multiple perspectives about the same nodes. Also, the user can isolate only the aspects that he chooses.

Now, ChatGPT, can you please... 😅

**Reply by dpolivaev** [reply]
I see quite many questions:

1. Do you really need clones?
2. Why do you want to keep the locations in the canvas between sessions?
3. I could easily add the option "single layout for the whole map" so that you can switch between layouts-per-node and layout-per-map. It is extension of @EdoFro suggestion. But I still do not see any use case. Which one do you have in mind?
4. Generally, if you have clones, they currently share all properties except for folding. Which properties should be individually configurable and for which use cases?

**Reply by euu2021** [reply]
> Do you really need clones?

Yes. In the past, I tried this setup with copies, and it doesn't work, because, as I'm working with the nodes in the canvas, I want to make changes in the nodes (adding text and children), and trying to locate the original node breaks the flow, making the experience frustrating. The beauty of this system is having the ideas getting enriched when put in a different context.

> Why do you want to keep the locations in the canvas between sessions?

Well, I see the canvas as a place for creative work, and creative work demands its own time. The idea is to leave it there and come to play with it from time to time.

Also, I like the idea of leaving it there after the project is finished, with the purpose of documentation. Probably, duplicating first, and breaking the clone link in one of the copies. It would be very cool to compare how the two copies get progressively more different as the knowledge base changes.


> Generally, if you have clones, they currently share all properties except for folding. Which properties should be individually configurable and for which use cases?

Well, I think the current behavior is fine. And let's not forget that, using conditional styles, the user can apply different formats for each clone. In this example, "ccc" are clones, but have different layouts:
![image](https://user-images.githubusercontent.com/77707706/226765023-0b39daeb-ce2f-4c37-8de9-60a876683c8e.png)

> I could easily add the option "single layout for the whole map" so that you can switch between layouts-per-node and layout-per-map. It is extension of @EdoFro suggestion. But I still do not see any use case. Which one do you have in mind?

While drafting the script idea that I posted two comments above, I noticed that this dichotomy "layouts-per-node and layout-per-map" is not enough for me. What I need is to open several different perspectives (like I can already open several different map views), because I can then use the same node on different projects (I'm thinking on one or more canvas per project) at the same time. So, my situation is different from Edo's. 

I think this script will solve my problem.

---
I will need to focus on an important project, so I won't be able to comment for some time.

**Comment by sifran-github** [comment]
Tested freeplane-1.11.1-pre21 with a few more layouts. I must say that its working well. I'm sure that there are more bugs, but in general the keyboard navigation is good, MUCH better than it was when i originally created this post.

One small bug that i found:
on "outline" layouts that goes up (only up, not down, for some reason), pressing UP skips a node sometimes.
![bug_keys](https://user-images.githubusercontent.com/5730355/227740604-2fe78f4d-d0ed-4134-bae7-7a47b63b5560.png)

Map:
[bug_keys.zip](https://github.com/freeplane/freeplane/files/11070043/bug_keys.zip)
